pkg/utils/log/log.go

```diff
@@ -0,0 +1,118 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package log
+
+import (
+	"flag"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/pkg/about"
+	"github.com/elastic/cloud-on-k8s/pkg/dev"
+	"github.com/go-logr/zapr"
+	pflag "github.com/spf13/pflag"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"k8s.io/klog"
+	crlog "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var (
+	verbosity       = flag.Int("log-verbosity", 0, "Verbosity level of logs (-2=Error, -1=Warn, 0=Info, >0=Debug)")
+	enableDebugLogs = flag.Bool("enable-debug-logs", false, "Enable debug logs")
+)
+
+// BindFlags attaches logging flags to the given flag set.
+func BindFlags(flags *pflag.FlagSet) {
+	flags.AddGoFlag(flag.Lookup("log-verbosity"))
+	flags.AddGoFlag(flag.Lookup("enable-debug-logs"))
+}
+
+// InitLogger initializes the global logger informed by the values of log-verbosity and enable-debug-logs flags.
+func InitLogger() {
+	setLogger(verbosity, enableDebugLogs)
+}
+
+// ChangeVerbosity replaces the global logger with a new logger set to the sepcified verbosity level.
+// Verbosity levels from 2 are custom levels that increase the verbosity as the value increases.
+// Standard levels are as follows:
+// level | Zap level | name
+// -------------------------
+//  1    | -1        | Debug
+//  0    |  0        | Info
+// -1    |  1        | Warn
+// -2    |  2        | Error
+func ChangeVerbosity(v int) {
+	debugLogs := false
+	setLogger(&v, &debugLogs)
+}
+
+func setLogger(v *int, debug *bool) {
+	level := determineLogLevel(v, debug)
+
+	// if the level is higher than 1 set the klog level to the same level
+	if level.Level() < zap.DebugLevel {
+		flagset := flag.NewFlagSet("", flag.ContinueOnError)
+		klog.InitFlags(flagset)
+		_ = flagset.Set("v", strconv.Itoa(int(level.Level())*-1))
+	}
+
+	var encoder zapcore.Encoder
+	var opts []zap.Option
+	if dev.Enabled {
+		encoderConf := zap.NewDevelopmentEncoderConfig()
+		encoderConf.EncodeLevel = zapcore.CapitalColorLevelEncoder
+		encoder = zapcore.NewConsoleEncoder(encoderConf)
+
+		opts = append(opts, zap.Development(), zap.AddStacktrace(zap.ErrorLevel))
+	} else {
+		encoderConf := zap.NewProductionEncoderConfig()
+		encoderConf.MessageKey = "message"
+		encoderConf.TimeKey = "@timestamp"
+		encoderConf.EncodeTime = zapcore.ISO8601TimeEncoder
+		encoder = zapcore.NewJSONEncoder(encoderConf)
+
+		opts = append(opts, zap.AddStacktrace(zap.WarnLevel))
```

Comment:
It seems a bit verbose to display stacktraces for warnings. Although I think that we never use the warning level in the code.

Reply:
I agree with you. I followed the [existing controller-runtime logger configuration](https://github.com/kubernetes-sigs/controller-runtime/blob/f1eaba5087d69cebb154c6a48193e6667f5b512c/pkg/runtime/log/log.go#L59)  but they seem to have found and [fixed the mistake in v0.2](https://github.com/kubernetes-sigs/controller-runtime/blob/699340657992a5ffe4d400a9f68330420fb4a689/pkg/log/zap/zap.go#L140) as well.